programs/clearing_house/src/math/amm.rs

```diff
@@ -1283,12 +1317,87 @@ pub fn calculate_max_base_asset_amount_fillable(amm: &AMM) -> ClearingHouseResul
     )
 }
 
+pub fn calculate_net_user_pnl(amm: &AMM, oracle_price: i128) -> ClearingHouseResult<i128> {
+    let net_user_base_asset_value = amm
+        .net_base_asset_amount
+        .checked_mul(oracle_price)
+        .ok_or_else(math_error!())?
+        .checked_div(AMM_RESERVE_PRECISION_I128 * cast_to_i128(PRICE_TO_QUOTE_PRECISION_RATIO)?)
+        .ok_or_else(math_error!())?;
+
+    let net_user_unrealized_pnl = net_user_base_asset_value
+        .checked_add(
+            amm.quote_asset_amount_long
+                .checked_add(amm.quote_asset_amount_short)
+                .ok_or_else(math_error!())?,
+        )
+        .ok_or_else(math_error!())?;
+
+    Ok(net_user_unrealized_pnl)
+}
+
+pub fn calculate_settlement_price(
+    amm: &AMM,
+    target_price: i128,
+    pnl_pool_amount: u128,
+) -> ClearingHouseResult<i128> {
+    // let long_unsettled_pnl = amm
+    //     .quote_asset_amount_long
+    //     .checked_sub(amm.quote_entry_amount_long)
+    //     .ok_or_else(math_error!())?;
+
+    // let short_unsettled_pnl = amm
+    //     .quote_asset_amount_short
+    //     .checked_sub(amm.quote_entry_amount_short)
+    //     .ok_or_else(math_error!())?;
+
+    // let total_unsettled_pnl = long_unsettled_pnl
+    //     .checked_add(short_unsettled_pnl)
+    //     .ok_or_else(math_error!())?;
+
+    if amm.net_base_asset_amount == 0 {
+        return Ok(target_price);
+    }
+
+    // net_baa * price + net_quote <= 0
+    // net_quote/net_baa <= -price
+
+    // net_user_unrealized_pnl negative = surplus in market
+    // net_user_unrealized_pnl positive = settlement price needs to differ from oracle
+    let best_settlement_price = -(amm
```

Comment:
gonna need you to explain this to me

Reply:
also we need to fix quote_asset_amount_long/short to be updated every time user.quote_asset_amount is updated